File: ruby/tournament/tournament.rb
class TeamStatus
  attr_reader :team, :wins, :losses, :draws, :matches

  def won
    @wins += 1
    @matches += 1
  end

  def lost
    @losses += 1
    @matches += 1
  end

  def tied
    @draws += 1
    @matches += 1
  end

  def points
    wins * 3 + draws
  end

  def <=>(other)
    point_diff = -(points <=> other.points)
    name_diff = team <=> other.team

    !point_diff.zero? ? point_diff : name_diff
  end

  private

  def initialize(team)
    @team = team
    @wins = 0
    @losses = 0
    @draws = 0
    @matches = 0
  end
end

class Tournament
  def self.tally(results)
    new(results.strip.lines.map(&:strip)).pretty
  end

  def pretty
    [header, *tournament_stats].join("\n") + "\n"
  end

  private

  attr_reader :teams

  def initialize(results)
    @teams = {}

    results.each do |line|
      team_a, team_b, result = line.split(';')
      ingest_result(team_a, team_b, result)
    end
  end

  def ingest_result(team_a, team_b, result)
    a_stats = ingest_team(team_a)
    b_stats = ingest_team(team_b)

    case result
    when 'win'
      a_stats.won
      b_stats.lost
    when 'loss'
      a_stats.lost
      b_stats.won
    when 'draw'
      a_stats.tied
      b_stats.tied
    end
  end

  def ingest_team(name)
    teams[name] ||= TeamStatus.new(name)
  end

  def header
    format_line('Team', 'MP', 'W', 'D', 'L', 'P')
  end

  def tournament_stats
    teams.each_value.sort.map do |stats|
      format_line(
        stats.team,
        stats.matches,
        stats.wins,
        stats.draws,
        stats.losses,
        stats.points
      )
    end
  end

  def format_line(team, mp, w, d, l, p)
    format('%-30s | %2s | %2s | %2s | %2s | %2s', team, mp, w, d, l, p)
  end
end

if $PROGRAM_NAME == __FILE__
  input = <<~INPUT
    Blithering Badgers;Allegoric Alaskans;win
  INPUT

  expected = <<~TALLY
    Team                           | MP |  W |  D |  L |  P
    Blithering Badgers             |  1 |  1 |  0 |  0 |  3
    Allegoric Alaskans             |  1 |  0 |  0 |  1 |  0
  TALLY

  puts Tournament.tally(input)
end

File: go/tournament/tournament.go
package tournament

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"sort"
	"strconv"
	"strings"
)

var (
	ErrBadInput = errors.New("bad input")
)

type Score struct {
	team   string
	played int
	won    int
	tied   int
	lost   int
	points int
}

func Tally(reader io.Reader, writer io.Writer) error {
	scoreByName, err := parse(reader)
	if err != nil {
		return err
	}

	sortedScores := sortScores(scoreByName)
	return writeResults(writer, sortedScores)
}

func parse(r io.Reader) (map[string]*Score, error) {
	scanner := bufio.NewScanner(r)
	tournament := map[string]*Score{}

	for scanner.Scan() {
		line := scanner.Text()
		if line == "" || line[0] == '#' {
			continue
		}

		entry := strings.Split(line, ";")
		if len(entry) != 3 {
			return nil, fmt.Errorf("line doesn't have 3 fields %q", line)
		}
		t1, t2, result := entry[0], entry[1], entry[2]

		if _, ok := tournament[t1]; !ok {
			tournament[t1] = &Score{team: t1}
		}

		if _, ok := tournament[t2]; !ok {
			tournament[t2] = &Score{team: t2}
		}

		tournament[t1].played++
		tournament[t2].played++

		switch result {
		case "win":
			tournament[t1].won++
			tournament[t1].points += 3

			tournament[t2].lost++
		case "draw":
			tournament[t1].tied++
			tournament[t1].points++

			tournament[t2].tied++
			tournament[t2].points++
		case "loss":
			tournament[t1].lost++

			tournament[t2].won++
			tournament[t2].points += 3
		default:
			return nil, fmt.Errorf("invalid result %q", result)
		}
	}

	return tournament, nil
}

func sortScores(scoresByName map[string]*Score) []*Score {
	scores := make([]*Score, 0, len(scoresByName))
	for _, score := range scoresByName {
		scores = append(scores, score)
	}

	sort.Slice(scores, func(i, j int) bool {
		switch {
		case scores[i].points > scores[j].points:
			return true
		case scores[i].points == scores[j].points:
			return scores[i].team < scores[j].team
		default:
			return false
		}
	})

	return scores
}

func writeResults(w io.Writer, results []*Score) error {
	err := writeLine(w, "Team", "MP", "W", "D", "L", "P")
	if err != nil {
		return err
	}

	for _, score := range results {
		err = writeLine(
			w,
			score.team,
			strconv.Itoa(score.played),
			strconv.Itoa(score.won),
			strconv.Itoa(score.tied),
			strconv.Itoa(score.lost),
			strconv.Itoa(score.points),
		)

		if err != nil {
			return err
		}
	}

	return nil
}

func writeLine(wr io.Writer, team, mp, w, d, l, p string) error {
	_, err := fmt.Fprintf(wr, "%-30s | %2s | %2s | %2s | %2s | %2s\n", team, mp, w, d, l, p)
	return err
}
